File: solutions/haiku-1/haiku/java/dijkstra.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Read N (number of nodes) and M (number of edges)
        int N = sc.nextInt();
        int M = sc.nextInt();

        // Build adjacency list for the graph
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            graph.add(new ArrayList<>());
        }

        // Read M edges
        for (int i = 0; i < M; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            int w = sc.nextInt();
            // Undirected graph: add edge in both directions
            graph.get(u).add(new int[]{v, w});
            graph.get(v).add(new int[]{u, w});
        }

        // Read start and target nodes
        int s = sc.nextInt();
        int t = sc.nextInt();

        // Special case: start equals target
        if (s == t) {
            System.out.println(0);
            return;
        }

        // Dijkstra's algorithm
        long[] dist = new long[N];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[s] = 0;

        // Priority queue: stores [distance, node]
        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        pq.add(new long[]{0, s});

        while (!pq.isEmpty()) {
            long[] curr = pq.poll();
            long d = curr[0];
            int u = (int) curr[1];

            // Skip if we've already found a better path
            if (d > dist[u]) continue;

            // Relax edges
            for (int[] edge : graph.get(u)) {
                int v = edge[0];
                int w = edge[1];

                if (dist[u] + w < dist[v]) {
                    dist[v] = dist[u] + w;
                    pq.add(new long[]{dist[v], v});
                }
            }
        }

        // Output result
        if (dist[t] == Long.MAX_VALUE) {
            System.out.println(-1);
        } else {
            System.out.println(dist[t]);
        }
    }
}

File: solutions/haiku-1/haiku/python/dijkstra.py
import heapq
from collections import defaultdict

def solve():
    # Read N, M
    n, m = map(int, input().split())

    # Build the graph
    graph = defaultdict(list)

    # Read edges
    for _ in range(m):
        u, v, w = map(int, input().split())
        graph[u].append((v, w))
        graph[v].append((u, w))  # undirected

    # Read s, t
    s, t = map(int, input().split())

    # If start equals target
    if s == t:
        return 0

    # Dijkstra's algorithm
    dist = [float('inf')] * n
    dist[s] = 0
    pq = [(0, s)]  # (distance, node)

    while pq:
        d, u = heapq.heappop(pq)

        if d > dist[u]:
            continue

        if u == t:
            return d

        for v, w in graph[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                heapq.heappush(pq, (dist[v], v))

    # If we reach here, there's no path
    return -1

print(solve())
